src/include/txn_inline.h

```diff
@@ -392,38 +392,97 @@ __wt_txn_op_delete_apply_prepare_state(WT_SESSION_IMPL *session, WT_REF *ref, bo
     WT_REF_UNLOCK(ref, previous_state);
 }
 
+/*
+ * __txn_op_delete_commit_verify_page_del_timestamp --
+ *     Verify the timestamps before setting it to the page delete structure.
+ */
+static WT_INLINE int
+__txn_op_delete_commit_verify_page_del_timestamp(
+  WT_SESSION_IMPL *session, WT_TXN_OP *op, bool locked)
+{
+    WT_ADDR_COPY addr;
+    WT_DECL_RET;
+    WT_PAGE_DELETED *page_del;
+    WT_REF *ref;
+    WT_REF_STATE previous_state;
+    WT_TXN *txn;
+    bool addr_found;
+
+    ref = op->u.ref;
+    txn = session->txn;
+    page_del = ref->page_del;
+
+    previous_state = WT_REF_GET_STATE(ref);
+    if (!locked)
+        WT_REF_LOCK(session, ref, &previous_state);
```

Comment:
We will be blocked here? Can we give up validation if we know we race here?

Reply:
No. We can't give up the validation. This can lead to other problems later if the timestamps are out-of-order.

Reply:
Another question is what can race with us here?

Reply:
Any normal operations/eviction that needs an exclusive access on the page can race with us.

Reply:
The page is on disk right? Do we also need to verify the page deleted structure even we have read back the deleted pages in memory?

Reply:
Yes, we have to check the validity of the truncate operation whether the fast truncated page was loaded into the memory or not?

Reply:
Should we trigger the check when the tombstone on the page is committed if the page is read back into memory?

Reply:
The tombstone is not yet committed. As part of the truncate operation commit, we are validating whether the commit timestamp of the truncate operation is not an out-of-order.

Reply:
So we can abort here if we cannot lock it?

Reply:
Sorry, I didn't get your point. We abort if we find the timestamps are out-of-order, we go to that situation only after we lock the page.

Reply:
If we cannot lock and directly return here, we should detect the out of order ts when we commit the tombstone from the deleted page, shouldn't we?

Reply:
This is the flow where we are committing, so we must wait to update the commit timestamp.